pczt/src/orchard.rs

```diff
@@ -0,0 +1,548 @@
+    /// For now, the Constructor sets both at the same time.
+    pub(crate) enc_ciphertext: Vec<u8>,
+    /// The encrypted note plaintext for the output.
+    ///
+    /// Encoded as a `Vec<u8>` because its length depends on the transaction version.
+    pub(crate) out_ciphertext: Vec<u8>,
+
+    /// The [raw encoding] of the Orchard payment address that will receive the output.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    ///
+    /// [raw encoding]: https://zips.z.cash/protocol/protocol.pdf#orchardpaymentaddrencoding
+    #[serde_as(as = "Option<[_; 43]>")]
+    #[getset(get = "pub")]
+    pub(crate) recipient: Option<[u8; 43]>,
+
+    /// The value of the output.
+    ///
+    /// This may be used by Signers to verify that the value matches `cv`, and to confirm
+    /// the values and change involved in the transaction.
+    ///
+    /// This exposes the value to all participants. For Signers who don't need this
+    /// information, we can drop the values and compress the rcvs into the bsk global.
+    #[getset(get = "pub")]
+    pub(crate) value: Option<u64>,
+
+    /// The seed randomness for the output.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover, instead of disclosing `shared_secret` to them.
+    #[getset(get = "pub")]
+    pub(crate) rseed: Option<[u8; 32]>,
+
+    /// The `ock` value used to encrypt `out_ciphertext`.
+    ///
+    /// This enables Signers to verify that `out_ciphertext` is correctly encrypted.
+    ///
+    /// This may be `None` if the Constructor added the output using an OVK policy of
+    /// "None", to make the output unrecoverable from the chain by the sender.
+    pub(crate) ock: Option<[u8; 32]>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the output.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// Proprietary fields related to the note being created.
+    #[getset(get = "pub")]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+impl Bundle {
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(
+        mut self,
+        other: Self,
+        self_global: &Global,
+        other_global: &Global,
+    ) -> Option<Self> {
+        // Destructure `other` to ensure we handle everything.
+        let Self {
+            mut actions,
+            flags,
+            value_sum,
+            anchor,
+            zkproof,
+            bsk,
+        } = other;
+
+        if self.flags != flags {
+            return None;
+        }
+
+        // If `bsk` is set on either bundle, the IO Finalizer has run, which means we
+        // cannot have differing numbers of actions, and the value sums must match.
+        match (self.bsk.as_mut(), bsk) {
+            (Some(lhs), Some(rhs)) if lhs != &rhs => return None,
+            (Some(_), _) | (_, Some(_))
```

Comment:
What's going on with the scrutinee here? In the first and fourth cases, the type appears to be `(Option<_>, Option<_>)` in the second and third cases, it appears to be `((Option<_>, _), (Option<_>, _))`? How does this even compile?

Reply:
Oops, missed the `|`.